File: backend/app/schemas/course.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

CourseStatus = Literal["active", "archived"]
CourseMemberRole = Literal["teacher", "student"]
CourseMemberStatus = Literal["active", "removed"]


class CourseCreate(BaseModel):
    name: str = Field(min_length=2, max_length=160, description="课程名称")
    description: str | None = Field(default=None, max_length=2000, description="课程简介")


class CourseUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=160, description="课程名称")
    description: str | None = Field(default=None, max_length=2000, description="课程简介")


class CourseJoinRequest(BaseModel):
    code: str = Field(min_length=4, max_length=64, description="课程加入码")


class CourseRead(BaseModel):
    id: int
    name: str
    description: str | None = None
    code: str
    invite_code: str
    teacher_id: int | None = None
    teacher_name: str | None = None
    student_count: int = 0
    current_user_role: CourseMemberRole | Literal["admin"] | None = None
    status: CourseStatus
    created_at: datetime
    updated_at: datetime


class CourseMemberProfileSnapshot(BaseModel):
    knowledge_score: float
    practice_score: float
    innovation_score: float
    exam_score: float
    efficiency_score: float
    quality_score: float
    learning_goal: str | None = None
    current_course: str | None = None
    weaknesses: list[str] = Field(default_factory=list)
    mastered_topics: list[str] = Field(default_factory=list)
    profile_summary: str | None = None
    updated_at: datetime


class CourseMemberRead(BaseModel):
    id: int
    course_id: int
    user_id: int
    username: str
    email: str
    full_name: str | None = None
    role: CourseMemberRole
    status: CourseMemberStatus
    profile: CourseMemberProfileSnapshot | None = None
    joined_at: datetime
    created_at: datetime
    updated_at: datetime


class CourseListResponse(BaseModel):
    items: list[CourseRead]
    total: int
    page: int
    page_size: int


class CourseMemberListResponse(BaseModel):
    items: list[CourseMemberRead]
    total: int
    page: int
    page_size: int


class CourseJoinResponse(BaseModel):
    course: CourseRead
    member: CourseMemberRead
    already_joined: bool = False

File: frontend/src/types/course.ts
export type CourseStatus = 'active' | 'archived'
export type CourseMemberRole = 'teacher' | 'student'
export type CourseMemberStatus = 'active' | 'removed'

export interface Course {
  id: number
  name: string
  description?: string | null
  code: string
  invite_code: string
  teacher_id?: number | null
  teacher_name?: string | null
  student_count: number
  current_user_role?: CourseMemberRole | 'admin' | null
  status: CourseStatus
  created_at: string
  updated_at: string
}

export interface CourseMember {
  id: number
  course_id: number
  user_id: number
  username: string
  email: string
  full_name?: string | null
  role: CourseMemberRole
  status: CourseMemberStatus
  profile?: CourseMemberProfileSnapshot | null
  joined_at: string
  created_at: string
  updated_at: string
}

export interface CourseMemberProfileSnapshot {
  knowledge_score: number
  practice_score: number
  innovation_score: number
  exam_score: number
  efficiency_score: number
  quality_score: number
  learning_goal?: string | null
  current_course?: string | null
  weaknesses: string[]
  mastered_topics: string[]
  profile_summary?: string | null
  updated_at: string
}

export interface CourseListResponse {
  items: Course[]
  total: number
  page: number
  page_size: number
}

export interface CourseMemberListResponse {
  items: CourseMember[]
  total: number
  page: number
  page_size: number
}

export interface CourseCreatePayload {
  name: string
  description?: string | null
}

export interface CourseUpdatePayload {
  name?: string | null
  description?: string | null
}

export interface CourseJoinPayload {
  code: string
}

export interface CourseJoinResponse {
  course: Course
  member: CourseMember
  already_joined: boolean
}
